src/python/pants/option/optionable.py

```diff
@@ -28,6 +44,130 @@ class Optionable(AbstractClass):
   deprecated_options_scope = None
   deprecated_options_scope_removal_version = None
 
+  @classmethod
+  def implementation_version(cls):
+    """
+    :API: public
+    """
+    return []
+
+  @classmethod
+  @memoized_method
+  def implementation_version_str(cls):
+    return '.'.join(['_'.join(map(str, x)) for x in cls.implementation_version()])
+
+  class CycleException(Exception):
+    """Thrown when a circular dependency is detected."""
+
+    def __init__(self, cycle):
+      message = 'Cycle detected:\n\t{}'.format(' ->\n\t'.join(
+          '{} scope: {}'.format(subsystem, subsystem.options_scope) for subsystem in cycle))
+      super(Optionable.CycleException, self).__init__(message)
+
+  _stable_name = None
+  @classmethod
+  def stable_name(cls):
```

Comment:
This concept is specific to tasks (as the docstring explains) so not sure it should move to Optionable. Instead, how about handling the stable_name in the overridden fingerprint property in Task?